src/lib/variable_length_ringbuffer/VariableLengthRingbuffer.cpp

```diff
@@ -0,0 +1,192 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2023 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+
+
+
+#include "VariableLengthRingbuffer.hpp"
+#include "assert.h"
+#include "string.h"
+
+
+VariableLengthRingbuffer::~VariableLengthRingbuffer()
+{
+	deallocate();
+}
+
+bool VariableLengthRingbuffer::allocate(size_t buffer_size)
+{
+	LockGuard lg{_mutex};
+
+	assert(_ringbuffer == nullptr);
+
+	_size = buffer_size;
+	_ringbuffer = new uint8_t[_size];
+	return _ringbuffer != nullptr;
```

Comment:
It's a peculiarity that our `new` implementation even can return nullptr, by the standard new should throw. I'd actually go as far and suggesting to just use plain malloc here.

Reply:
I thought new without exceptions enabled would return `nullptr`. Is that wrong?

Reply:
In our implementation it does afaik, but behaviour of `new` on systems without exceptions is technically undefined. I think most implementations call `abort` when it happens. There is a `new(std::nothrow)` from the standard library that is allowed to return null, but yeah, it's all a bit messy

Reply:
Oh right, I have heard of that one. I'll change to malloc.

Reply:
If it's a real issue we need to pursue a systemic fix, nearly every module is already doing this.

Reply:
Ok let's leave it as is.